### PR TITLE
Refactor MoE routing with vectorized dispatch and load balancing

### PR DESCRIPTION
**Pull request by dansasser**
## Summary
- refactor `MoELayer` to compute routing maps once, add capacity-aware token dispatch, and expose a gradient-carrying load balancing loss
- store recent assignment counts and expose a helper so callers/tests can inspect expert utilization
- expand the phase 2 optimization tests with a reference MoE implementation to verify numerical parity and benchmark vectorized throughput

## Testing
- `pytest 'SIM-ONE Training/test_phase2_optimizations.py'` *(fails: missing torch dependency in the execution environment)*
- `pip install torch --index-url https://download.pytorch.org/whl/cpu` *(fails: proxy 403 when accessing download.pytorch.org)*
- `pip install https://download.pytorch.org/whl/cpu/torch-2.3.0%2Bcpu-cp312-cp312-linux_x86_64.whl` *(fails: proxy 403 when accessing download.pytorch.org)*
- `pip install torch==2.1.0` *(fails: wheel unavailable for Python 3.12)*

------
https://chatgpt.com/codex/tasks/task_e_68cfd214e7bc832ea6d8d23cb747452e

## Summary by CodeRabbit

- New Features
  - Capacity-aware routing with configurable capacity and guaranteed token assignment; new observability APIs for load-balancing metrics and stats reset.
- Performance
  - Vectorized, batched expert dispatch for faster forwards while matching prior correctness.
- Reliability
  - Handles empty inputs and overloads gracefully; enforces capacity limits with fallback routing.
- Tests
  - Added parity and speed benchmarks, expanded assertions for capacity, gradients, and stats reset.

### CODE REVIEW

File: SIM-ONE Training/simone_transformer/modern_layers.py [hunk 1]
```diff
@@ -83,111 +83,159 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            sorted_experts = expanded_experts[sorted_order]
+            sorted_cumsum = F.one_hot(sorted_experts, num_classes=self.num_experts).cumsum(dim=0)
+            expert_position = sorted_cumsum[torch.arange(sorted_cumsum.size(0), device=x_flat.device), sorted_experts] - 1
+            keep_sorted = expert_position < capacity
+            keep_mask = torch.zeros_like(expanded_weights, dtype=torch.bool)
+            keep_mask[sorted_order] = keep_sorted
+        else:
+            capacity = None
+            keep_mask = torch.ones_like(expanded_weights, dtype=torch.bool)
+
```
[the remark this answers is not in the document]
Comment (dansasser) [reply SIM-ONE Training/simone_transformer/modern_layers.py:160]: @codex
In SIM-ONE Training/simone_transformer/modern_layers.py around lines 146-160,
the current capacity masking uses F.one_hot plus cumsum which allocates an
O(num_tokens*k * num_experts) tensor; replace that with a run-length approach
that computes per-expert positions from sorted_experts without one_hot: detect
group boundaries in sorted_experts (compare adjacent entries), get group start
indices, compute run lengths from those starts, expand start indices by run
lengths to align with the sorted positions, subtract the expanded starts from
the global positions (0..M-1) to get zero-based position within each expert
group, then compare that position to capacity to form keep_sorted and map it
back via sorted_order to produce keep_mask; keep variable names (sorted_order,
sorted_experts, keep_sorted, keep_mask) and ensure tensors are on the correct
device and dtype to avoid extra allocations.
